### PR TITLE
sg/msp: add cloud deploy pubsub notifications

### PR DESCRIPTION
**Pull request by jac**
Adds notifications for success and failure of cloud deploy rollouts by sending pubsub events to an endpoint.

Adds the `MSP Rollouts` bot to provisioned #alerts-<service>-<env> channels so it can send notifications.
It adds it to all service channels even if they're not using rollouts which I think is fine

We get the endpoint to send notifications to from a secret in `sourcegraph-secrets`

Closes CORE-30

## Test plan

locally tested

### CODE REVIEW

File: dev/managedservicesplatform/googlesecretsmanager/googlesecretsmanager.go [hunk 1]
```diff
@@ -45,4 +43,6 @@ const (
 	// for configuring Cloudflare TLS.
 	SecretSourcegraphWildcardKey  = "SOURCEGRAPH_WILDCARD_KEY"
 	SecretSourcegraphWildcardCert = "SOURCEGRAPH_WILDCARD_CERT"
+	// SecretMSPDeployNotifEndpoint is the endpoint that MSP uses for cloud deploy push notifications
```
Comment (unknwon) [inline dev/managedservicesplatform/googlesecretsmanager/googlesecretsmanager.go:46]:
```suggestion
	// SecretMSPDeployNotifEndpoint is the endpoint that MSP uses for cloud deploy push notifications.
```

File: dev/managedservicesplatform/stacks/cloudrun/cloudrun.go [hunk 2]
```diff
@@ -461,6 +464,35 @@ func NewStack(stacks *stack.Set, vars Variables) (crossStackOutput *CrossStackOu
 			skaffoldBucketName: skaffoldBucket.Name(),
 			pipelineBucketName: pipelineBucket.Name(),
 		})
+
+		// Enable pubsub topics to receive notifications for cloud deploy events
```
[the remark this answers is not in the document]
Comment (jac) [reply dev/managedservicesplatform/stacks/cloudrun/cloudrun.go:468]: hmm I don't think `Grab` is the right term here.
The topics don't already exist that we fetch/grab them. Maybe Create?
```suggestion
		// Create the Pub/Sub topic that receives notifications for Cloud Deploy events,
```

[the remark this answers is not in the document]
Comment (unknwon) [reply dev/managedservicesplatform/stacks/cloudrun/cloudrun.go:468]: My understanding is that the topic exists already (created by Cloud Deploy?), so we are not creating at least 🤔 or is that not the case?

[the remark this answers is not in the document]
Comment (jac) [reply dev/managedservicesplatform/stacks/cloudrun/cloudrun.go:468]: There are defined topic names that cloud deploy will send events to _if_ the topics exist.
We have to create the topics though. They are not created by Cloud Deploy

[the remark this answers is not in the document]
Comment (unknwon) [reply dev/managedservicesplatform/stacks/cloudrun/cloudrun.go:468]: Oh I see, in that case, "Create" is the right verb 👍

File: dev/managedservicesplatform/stacks/monitoring/monitoring.go [hunk 3]
```diff
@@ -122,6 +122,9 @@ const StackName = "monitoring"
 
 const sharedAlertsSlackChannel = "#alerts-msp"
 
+// mspRolloutsSlackUserID  is the user ID of MSP Rollouts Slack bot.
+const mspRolloutsSlackUserID string = "U072KKSCLSJ"
```
Comment (bobheadxi) [inline dev/managedservicesplatform/stacks/monitoring/monitoring.go:126]:
```suggestion
// mspRolloutsBotSlackUserID  is the user ID of MSP Rollouts Slack bot.
const mspRolloutsBotSlackUserID string = "U072KKSCLSJ"
```

File: dev/managedservicesplatform/stacks/monitoring/monitoring.go [hunk 4]
```diff
@@ -267,10 +270,12 @@ func NewStack(stacks *stack.Set, vars Variables) (*CrossStackOutput, error) {
...
-				Name:      pointers.Ptr(strings.TrimPrefix(channel.Name, "#")),
-				Topic:     description,
-				Purpose:   description,
-				IsPrivate: pointers.Ptr(false),
+				Name:                           pointers.Ptr(strings.TrimPrefix(channel.Name, "#")),
+				Topic:                          description,
+				Purpose:                        description,
+				IsPrivate:                      pointers.Ptr(false),
+				PermanentMembers:               pointers.Ptr(pointers.Slice([]string{mspRolloutsSlackUserID})),
+				ActionOnUpdatePermanentMembers: pointers.Ptr("none"),
```
Comment (bobheadxi) [inline dev/managedservicesplatform/stacks/monitoring/monitoring.go:278]:
```suggestion
				// Do not kick out other users in the channel
				ActionOnUpdatePermanentMembers: pointers.Ptr("none"),
```

Comment (bobheadxi) [reply dev/managedservicesplatform/stacks/monitoring/monitoring.go:278]: helpful to leave a doc comment if the config's name doesn't make it immediately clear what the config for

File: dev/managedservicesplatform/stacks/cloudrun/cloudrun.go [hunk 5]
```diff
@@ -461,6 +463,35 @@ func NewStack(stacks *stack.Set, vars Variables) (crossStackOutput *CrossStackOu
...
 		})
+
+		// Create the Pub/Sub topic that receives notifications for Cloud Deploy events,
+		// see https://cloud.google.com/deploy/docs/subscribe-deploy-notifications#available_topics for topic info.
+		topic := pubsubtopic.NewPubsubTopic(stack, id.TerraformID("clouddeploy-operations-topic"), &pubsubtopic.PubsubTopicConfig{
+			Name: pointers.Ptr("clouddeploy-operations"),
+		})
+
+		// Get cloud-relay endpoint from GSM.
+		endpoint := gsmsecret.Get(stack, id.Group("clouddeploy-endpoint"), gsmsecret.DataConfig{
```
Comment (bobheadxi) [inline dev/managedservicesplatform/stacks/cloudrun/cloudrun.go:474]: Should this be `cloudrelay-endpoint`?

**Review by bobheadxi — APPROVED** [review]
Minor nits :) Exciting!

File: dev/managedservicesplatform/stacks/cloudrun/cloudrun.go (hunk 5 above)
Comment (jac) [reply dev/managedservicesplatform/stacks/cloudrun/cloudrun.go:474]: ah yeah, mb

**Comment by bobheadxi** [comment]
Also @jac , reminder to tag the relevant Linear issue - if it doesn't close it, you can use `related to CORE-...`, but if you have multiple PRs closing it you can mark them all as `closes` and Linear won't close the issue until they all merge :)

**Pull request merged**